File: data/validator.go
package data

//go:generate counterfeiter -o ./fakeValidator.go --fake-name fakeValidator ./ validator

import (
	"errors"
	"github.com/chrisdostert/gojsonschema"
	"github.com/golang-interfaces/ios"
	"github.com/opspec-io/sdk-golang/model"
)

type validator interface {
	Validate(
		value *model.Value,
		param *model.Param,
	) (errors []error)
}

func newValidator() validator {
	// register custom format checkers
	gojsonschema.FormatCheckers.Add("docker-image-ref", DockerImageRefFormatChecker{})
	gojsonschema.FormatCheckers.Add("integer", IntegerFormatChecker{})
	gojsonschema.FormatCheckers.Add("semver", SemVerFormatChecker{})

	return _validator{
		os:      ios.New(),
		coercer: newCoercer(),
	}
}

type _validator struct {
	coercer coercer
	os      ios.IOS
}

// Validate validates a value against a parameter
// note: param defaults aren't considered
func (this _validator) Validate(
	value *model.Value,
	param *model.Param,
) (errs []error) {
	if nil == param {
		return []error{errors.New("param required")}
	}

	switch {
	case nil != param.Dir:
		errs = this.validateDir(value)
	case nil != param.File:
		errs = this.validateFile(value)
	case nil != param.String:
		errs = this.validateString(value, param.String.Constraints)
	case nil != param.Number:
		errs = this.validateNumber(value, param.Number.Constraints)
	case nil != param.Object:
		errs = this.validateObject(value, param.Object.Constraints)
	case nil != param.Socket:
		errs = this.validateSocket(value)
	}
	return
}

File: lib/op/interpreter/opcall/params/param/validator.js
require('../../../../../model/data')
require('../../../../../model/param')
const arrayValidator = require('./array/validator')
const booleanValidator = require('./boolean/validator')
const dirValidator = require('./dir/validator')
const fileValidator = require('./file/validator')
const numberValidator = require('./number/validator')
const objectValidator = require('./object/validator')
const socketValidator = require('./socket/validator')
const stringValidator = require('./string/validator')

class Validator {
  /**
   * Validates value against param
   * note: param defaults aren't considered
   * @param {Value} value
   * @param {Param} param
   * @returns {Array<String>}
   */
  validate (
    value,
    param
  ) {
    if (!param) {
      throw new Error('param required')
    }

    if (param.array) {
      return arrayValidator.validate(
        value.array,
        param.array.constraints
      )
    }

    if (param.boolean) {
      return booleanValidator.validate(
        value.boolean
      )
    }

    if (param.dir) {
      return dirValidator.validate(
        value.dir
      )
    }

    if (param.file) {
      return fileValidator.validate(
        value.file
      )
    }

    if (param.number) {
      return numberValidator.validate(
        value.number,
        param.number.constraints
      )
    }

    if (param.object) {
      return objectValidator.validate(
        value.object,
        param.object.constraints
      )
    }

    if (param.socket) {
      return socketValidator.validate(
        value.socket
      )
    }

    if (param.string) {
      return stringValidator.validate(
        value.string,
        param.string.constraints
      )
    }
  }
}

// export as singleton
module.exports = new Validator()
